#! /bin/sh -e

# DP: Fix Ubuntu #12744, bug in StreamSerializer, seen with eclipse-3.1

dir=
if [ $# -eq 3 -a "$2" = '-d' ]; then
    pdir="-d $3"
    dir="$3/"
elif [ $# -ne 1 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch)
        patch $pdir -f --no-backup-if-mismatch -p0 < $0
        ;;
    -unpatch)
        patch $pdir -f --no-backup-if-mismatch -R -p0 < $0
        ;;
    *)
        echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
        exit 1
esac
exit 0

diff -urN libjava/gnu/xml/transform/StreamSerializer.java libjava/classpath/gnu/xml/transform/StreamSerializer.java
--- libjava/gnu/xml/transform/StreamSerializer.java	2005-09-24 02:23:36.000000000 +0200
+++ libjava/gnu/xml/transform/StreamSerializer.java	2005-07-16 02:33:40.000000000 +0200
@@ -627,12 +627,20 @@
         text = buf.toString();
       }
     ByteBuffer encoded = encoder.encode(CharBuffer.wrap(text));
+    int len = encoded.limit() - encoded.position();
     if (encoded.hasArray())
       {
-        return encoded.array();
+        byte[] ret = encoded.array();
+        if (ret.length > len)
+          {
+            // Why?
+            byte[] ret2 = new byte[len];
+            System.arraycopy(ret, 0, ret2, 0, len);
+            ret = ret2;
+          }
+        return ret;
       }
     encoded.flip();
-    int len = encoded.limit() - encoded.position();
     byte[] ret = new byte[len];
     encoded.get(ret, 0, len);
     return ret;
